js/src/specs.js

```diff
@@ -4,155 +4,180 @@
+   */
   database(): Database {
-    if (this.scheme === 'mem') {
-      return new Database(new BatchStoreAdaptor(new MemoryStore()));
-    }
-    if (this.scheme === 'http') {
-      return new Database(new HttpBatchStore(`${this.scheme}:${this.path}`));
+    switch (this.scheme) {
+      case 'mem':
+        return new Database(new BatchStoreAdaptor(notNull(this._ms)));
+      case 'http':
+      case 'https':
+        return new Database(new HttpBatchStore(`${this.scheme}:${this.path}`));
+      default:
+        throw new Error('Unreached');
     }
-    throw new Error('Unreached');
   }
 }
 
-// A parsed specification for the location of a Noms dataset.
-// For example: 'mem:photos' or 'https://ds.noms.io/aa/music:funk'
-//
-// See "spelling datasets" for details on supported syntaxes:
-// https://docs.google.com/document/d/1QgKcRS304llwU0ECahKtn8lGBFmT5zXzWr-5tah1S_4/edit
+/**
+ * A parsed specification for the location of a Noms dataset.
+ * For example: 'mem::photos' or 'https://demo.noms.io/cli-tour::sf-crime'.
+ *
+ * See https://github.com/attic-labs/noms/blob/master/doc/spelling.md.
+ */
 export class DatasetSpec {
   database: DatabaseSpec;
   name: string;
 
-  // Returns a parsed spec, or null if the spec was invalid.
-  static parse(spec: string): ?DatasetSpec {
-    const match = spec.match(/^(.+)\:\:([a-zA-Z0-9\-_/]+)$/);
-    if (!match) {
-      return null;
+  /**
+   * Returns `spec` parsed as a DatasetSpec. Throws a `SyntaxError` if `spec` isn't valid.
+   */
+  static parse(spec: string): DatasetSpec {
+    const [database, name] = splitAndParseDatabaseSpec(spec);
+    if (!datasetRe.test(name)) {
+      throw new SyntaxError(`Invalid dataset ${name}, must match ${datasetRe.source}`);
     }
-    const [, dbSpec, dsName] = match;
-    const database = DatabaseSpec.parse(dbSpec);
-    if (!database) {
-      return null;
-    }
-    return new this(database, dsName);
+    return new DatasetSpec(database, name);
   }
 
   constructor(database: DatabaseSpec, name: string) {
     this.database = database;
     this.name = name;
   }
 
-  // Returns a new DataSet based on the parsed spec.
+  /**
+   * Returns a new Dataset based on this DatasetSpec.
+   */
   dataset(): Dataset {
     return new Dataset(this.database.database(), this.name);
   }
 
-  // Returns the value at the HEAD of this dataset, if any, or null otherwise.
-  value(): Promise<any> {
-    // Hm. Calling dataset() creates a Database that we then toss into the ether, which means we
-    // can't call close() on it. Ideally, we'd fix that.
-    return this.dataset().head()
-      .then(commit => commit && commit.value);
+  /**
+   * Returns a tuple of the database backed by this dataset, and value at the HEAD of this
+   * dataset. If this dataset doesn't have a HEAD, the value will be `null`.
+   *
+   * The caller should always call `close()` when done.
+   */
+  value(): Promise<[Database, Value|null]> {
```

Comment:
It's a breaking change anyway (return type changed) so I *can* change it without any more breakage. Let me know if you want that.